### PR TITLE
Overhaul GRPO training loop around exact-only rewards

### PR DESCRIPTION
**Pull request by workofart**
## Summary
Two commits:

1. **`load_checkpoint_metadata`** (`autograd/tools/model.py`) — reads bookkeeping values (step count, `model_init_kwargs`, …) from the JSON sidecar without opening the NPZ. GRPO startup previously ran a full `load_checkpoint` just to read `model_init_kwargs`, then the trainer deserialized the same checkpoint again for the weights. Fails fast if an array hides outside the skipped state-dict keys; tested by deleting the NPZ before loading.
2. **GRPO overhaul** (`examples/grpo.py`, `examples/gpt_2.py`, `test/test_grpo.py`):
   - **Exact-only reward**: 1.0 only for a fully format-valid completion (`<think>/<answer>` closed by the turn separator) whose answer matches exactly; everything else 0.0. Settled by the accuracy experiment log (E24: strict parser is not hiding accuracy; E33: no surface signal separates correct rollouts) — the old shaped per-tag rewards (1.4/0.2/0.3) rewarded incoherent arithmetic without lifting pass@1.
   - **Rollouts use the SFT special-token chat format** (R1-style system prompt, `<|USER|>`/`<|ASSISTANT|>` markers), so GRPO starts from exactly what SFT taught.
   - **Sparse output head**: `GPT2.forward` optionally projects only selected token rows; prompt/pad rows have zero GRPO weight, so their final layernorm + tied vocab projection are wasted. Isolated 64-token train-step probe: 0.472s → 0.378s (~1.25×, `notes/grpo_optimization.md`). A new test proves the selected-row loss matches the full-logits loss on the same batch.
   - **Batch-local padding** in `GRPOCollator` (pad to the longest row, not `max_tokens`), zero-advantage **step skipping** (uniform rewards carry no gradient signal; skipping also avoids Adam momentum drift from a zero gradient), and rollout-based **validation callbacks** (pass@1 / any@k / majority@k on a seen-train slice and a held-out GSM8K slice).

## Evidence honesty
The loop design is what the experiment log converged on, but held-out GSM8K pass@1 stayed ≈0 with the 124M base model in those experiments — pre-GRPO pass@1 (~0.016) is the binding constraint, since GRPO concentrates existing capability rather than creating it. The default config comments this explicitly and points at swapping in an SFT checkpoint.

## Excluded (kernel/experimental bucket)
KV-cache generation paths, `warmup_kv`, fused GRPO loss kernels, and the `gpt_2_llama` example stay on the experimental branch. The fused scale+clip optimizer flag is unnecessary here — it became the default in #88.

## Test plan
- New: sparse-vs-full loss equivalence (nonzero loss, rtol 1e-5), metadata loader round-trip with NPZ deleted, array-outside-skip-keys rejection.
- Ported: exact-only reward cases (wrong answer / missing think / missing EOS / unformatted all 0.0), dynamic-padding collator shapes, config contract (`temperature > 0`, `top_k=None`), rollout context-window rejection.
- Full suite: 571 passed, 11 skipped. Ruff lint + format clean.

### CODE REVIEW

File: autograd/tools/model.py [hunk 1]
```diff
@@ -258,3 +258,60 @@ def _deserialize(meta: SerializedMeta, data: Dict[str, Any]) -> Any:
...
+    if t in ("scalar", "raw"):
+        return meta["value"]
+
+    raise ValueError(
+        f"Checkpoint metadata cannot contain array data (got _type={t!r}); "
+        "use load_checkpoint to deserialize arrays from the NPZ file"
+    )
+
+
+def load_checkpoint_metadata(
```
Comment (workofart) [inline autograd/tools/model.py:290]: we can avoid the memory spike and instead just load metadata from the checkpoint
